### PR TITLE
Summon Pitbulls

### PR DESCRIPTION
**Pull request by RealestEstate**
A wizard spell that summons 3 beautiful pitbulls per casting to cuddle and play with the crew.
Up to a maximum of 15 living at a time.

woofer

sprites courtesy of Drawstuff

Edit:
-Added a Treachery mechanic to pitbulls so they may occasionally bite their masters

:cl:
 * rscadd: Adds Summon Pitbulls spell.

### CODE REVIEW

File: code/modules/spells/aoe_turf/conjure/pitbulls.dm [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+#define MAX_PITBULLS 15
+
+/spell/aoe_turf/conjure/pitbull/cast_check(skipcharge = 0,mob/user = usr)
+	if (pitbulls_count_by_wizards[user] > MAX_PITBULLS) // We summoned too many pitbulls :(
+		to_chat(user, "<span class = 'warning'>We've brought forth too many innocent pitbulls into this plane.</span>")
+		return FALSE
+	return ..()
+
+
+/obj/testing
```
Comment (ShiftyRail) [inline code/modules/spells/aoe_turf/conjure/pitbulls.dm]: :thinking:

Comment (RealestEstate) [reply code/modules/spells/aoe_turf/conjure/pitbulls.dm]: oobs I forget

**Comment by SonixApache** [comment]
dumb meme spell

File: code/modules/spells/aoe_turf/conjure/pitbulls.dm [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+/spell/aoe_turf/conjure/pitbull/cast_check(skipcharge = 0,mob/user = usr)
+	if (pitbulls_count_by_wizards[user] > MAX_PITBULLS) // We summoned too many pitbulls :(
+		to_chat(user, "<span class = 'warning'>We've brought forth too many innocent pitbulls into this plane.</span>")
+		return FALSE
+	return ..()
+
+
+/spell/aoe_turf/conjure/pitbull/summon_object(var/type, var/location)
+	..()
+	pitbulls_count_by_wizards[holder]++
```
Comment (DamianX) [inline code/modules/spells/aoe_turf/conjure/pitbulls.dm]: the wizard is never removed from this global list, so it will never be properly deleted.
either clear it manually or use weak references through `\ref`

File: code/modules/mob/living/simple_animal/hostile/pitbull.dm [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		"Good with kids")
+
+/mob/living/simple_animal/hostile/pitbull/death()
+	..(TRUE)
+	var/mob/my_wiz = pitbulls[src]
+	pitbulls_count_by_wizards[my_wiz]--
+	pitbulls[src] = null
+	pitbulls -= src
+
+/mob/living/simple_animal/hostile/pitbull/ListTargets()
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/pitbull.dm:42]: you should be able to set the pitbull's `faction` var to the `\ref[]` of the wizard instead of doing any of this

File: code/modules/mob/living/simple_animal/hostile/pitbull.dm [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+		"Blame the owner not the breed!",
+		"My dog would never do that.",
+		"He's a big baby at heart.",
+		"Man's best friend.",
+		"Over 4 million pitbulls did not kill or hurt anyone today. Stop the myth.",
+		"Good with kids")
+
+/mob/living/simple_animal/hostile/pitbull/death()
+	..(TRUE)
+	var/mob/my_wiz = pitbulls[src]
```
Comment (DamianX) [inline code/modules/mob/living/simple_animal/hostile/pitbull.dm]: it seems nothing ever adds an entry to this `pitbulls` list, am I missing something?

**Comment by boy2mantwicethefam** [comment]
kek

**Comment by DrSnips** [comment]
Kind of powercreeps the butterfly knife but I doubt anyone cares.

File: code/modules/mob/living/simple_animal/hostile/pitbull.dm (hunk 3 above)
Comment (Kurfursten) [reply code/modules/mob/living/simple_animal/hostile/pitbull.dm:42]: I think he wants the pitbulls to not be on a faction with the wizard

you know, "heroic pitbull travels 200 tiles across station to maul master"

Comment (DrSnips) [reply code/modules/mob/living/simple_animal/hostile/pitbull.dm:42]: They are currently bound to the wizard faction already so it shouldn't be attacking wizards or other pitbulls by default. Mobs of the same faction don't target each other.

Comment (RealestEstate) [reply code/modules/mob/living/simple_animal/hostile/pitbull.dm:42]: I was trying to exclude the master from being a target but I didn’t quite do it right.
maybe what I can do is have a life check to specifically turn on the master kind of like how goats get aggressive

File: code/modules/mob/living/simple_animal/hostile/pitbull.dm [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+/mob/living/simple_animal/hostile/pitbull/New()
+	..()
+	desc = pick(
+		"There is no such thing as a bad dog, only bad owners.",
+		"Blame the owner not the breed!",
+		"My dog would never do that.",
+		"He's a big baby at heart.",
+		"Man's best friend.",
+		"Over 4 million pitbulls did not kill or hurt anyone today. Stop the myth.",
+		"Good with kids")
```
Comment (DrSnips) [inline code/modules/mob/living/simple_animal/hostile/pitbull.dm]: .

**Comment by RealestEstate** [comment]
@DrSnips
yeah that’s kind of unfortunate I totally forgot about it

File: code/modules/spells/aoe_turf/conjure/pitbulls.dm [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+	var/list/turf/locs = new
+
+	for(var/direction in alldirs) //looking for pitbull spawns
+		if(locs.len >= 3) //we found 3 locations and thats all we need
+			break
+		var/turf/T = get_step(user, direction) //getting a loc in that direction
+		if(AStar(user.loc, T, /turf/proc/AdjacentTurfs, /turf/proc/Distance, 1)) // if a path exists, so no dense objects in the way its valid salid
+			locs += T
+		else
```
Comment (DamianX) [inline code/modules/spells/aoe_turf/conjure/pitbulls.dm]: why is this here?

**Comment by Hoshi-chan** [comment]
Baby's day out when?

File: code/modules/mob/living/simple_animal/hostile/pitbull.dm (hunk 3 above)
Comment (DamianX) [reply code/modules/mob/living/simple_animal/hostile/pitbull.dm:42]: I just told you how to exclude the master from being a target

Comment (RealestEstate) [reply code/modules/mob/living/simple_animal/hostile/pitbull.dm:42]: yeah I saw and I plan to implement it. I was trying to answer kurf's question while also adding the idea of a life check that makes the pitbull turn specifically on the owner.

**Comment by KingofHONK** [comment]
you're obviously an ignorant piece of shit that has no idea what you're talking about. I've never met a single pitbull owner that acted like this in my entire life and just about everyone I've ever met either own pit bulls or has in the past... and no we don't live in a poor town, not by a long stretch.

oh yeah and just to add one more tidbit of information that just shows how fucking retarded you are - of the thousands of pitbulls I've known throughout my life, only two of them were aggressive at all and they were trying to be attack dogs.. the rest were nothing but huge babies. Contrary to that I can't even count the number of mean labs and other breeds I've met on two hands..

so this just goes to show that your extensive knowledge of pitbull aggression comes from TV and the internet and not from actual personal experience with responsible dog owner

**Comment by ArchSunder** [comment]
this is a huge nerf to baby's day out mode

**Comment by yclatious** [comment]
Realest is saving /verygay/ station, one powerful pr at a time. God bless.

**Comment by Killette2** [comment]
> Kind of powercreeps the butterfly knife but I doubt anyone cares.

@DrSnips butterfly knife still does clone damage which keeps it relevant for if you wanted to meme with it. It's a hard to replicate niche, and consider that 1 butterfly does more damage than 3 pitbulls albeit with the health of 1.

**Comment by ghost** [comment]
make upgrading the spell summon SMASHED & SLAMMED TOADLINE MAFIA BOSSES

File: code/modules/spells/aoe_turf/conjure/pitbulls.dm (hunk 2 above)
Comment (RealestEstate) [reply code/modules/spells/aoe_turf/conjure/pitbulls.dm]: @DamianX I might need some help implementing the weak reference, I have no idea how to even start.

Comment (DamianX) [reply code/modules/spells/aoe_turf/conjure/pitbulls.dm]: Actually, weak references are stupid.
Just do
```
if(pitbull_count_by_wizard[my_wiz] == 0)
    pitbulls_count_by_wizard -= my_wiz
```
or something like that after you decreased the count in `pitbull/death()`

Comment (RealestEstate) [reply code/modules/spells/aoe_turf/conjure/pitbulls.dm]: @DamianX 
I already have `pitbulls_count_by_wizards[my_wiz]--` in `pitbull/death()`, is that sufficient?

Comment (RealestEstate) [reply code/modules/spells/aoe_turf/conjure/pitbulls.dm]: oh wait I misunderstood, yeah I gotta delete the list when it's empty ignore that last comment

Comment (DamianX) [reply code/modules/spells/aoe_turf/conjure/pitbulls.dm]: You not-so-much have to delete the list, it's more that you have to remove the reference to the wizard from the list, at least.

**Comment by tuypo1** [comment]
Aren't the smashed and slamed dogs pugs though

**Comment by crudequip** [comment]
no toadline are inbred pitbulls. still abominations, but without a lot of the problems that pugs have (in exchange for joint and cartilage damage by age 4)

**Comment by RealestEstate** [comment]
need fixing and conflict tags should be resolved, ready for final reviews

File: code/modules/mob/living/simple_animal/hostile/pitbull.dm [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+/mob/living/simple_animal/hostile/pitbull/proc/Calmdown()
+	faction = faction_original
+	friends = friends_temp.Copy()
+	if(target && ismob(target))
+		var/mob/M = target
+		if((M.faction == faction) || (M in src.friends))//stop chasing our friend when we calm down
+			LoseTarget()
+	treacherous = FALSE
+
+//some dude named yclat asked for this, I'm so fucking sorry aaaaahhh
```
Comment (Kurfursten) [inline code/modules/mob/living/simple_animal/hostile/pitbull.dm]: having a meme instakill effect is already something to be leery about but putting meme comments in the code - especially those that reference people by name - should be avoided outright
